[WebXR][OpenXR] Content not shown when asking for canvas without alpha layer

svillar opened the pull request:
#### 60a2c8c774dc80e2ef5b695f8eb365084c43ba0a
<pre>
[WebXR][OpenXR] Content not shown when asking for canvas without alpha layer
<a href="https://bugs.webkit.org/show_bug.cgi?id=299737">https://bugs.webkit.org/show_bug.cgi?id=299737</a>

Reviewed by Dan Glastonbury.

Content was not rendered on screen because we were trying to
blit buffers with different pixel formats in case an alpha channel
was not requested. We were creating and OpenXR texture with GL_RGB
but WebGL&apos;s opaque framebuffer unconditionally uses GL_RGBA.

It is not possible to make the opaque framebuffer to use
RGB instead because it has always to import a texture with an
alpha channel because ANGLE does not support creating an
external texture from a framebuffer with RGB format
(see Source/ThirdParty/ANGLE/src/common/linux/dma_buf_utils.cpp).

So the only solution is to make OpenXR always create swapchains
with alpha channel and then when exporting the texture just
ignore it (by using DRM_FORMAT_XRGBA8888). This way WebGL
and OpenXR will use the same format and blitting will work.

* Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.cpp:
(WebKit::OpenXRLayer::exportOpenXRTextureGBM):
* Source/WebKit/UIProcess/XR/openxr/OpenXRSwapchain.cpp:
(WebKit::OpenXRSwapchain::create):
(WebKit::OpenXRSwapchain::OpenXRSwapchain):
* Source/WebKit/UIProcess/XR/openxr/OpenXRSwapchain.h:
(WebKit::OpenXRSwapchain::hasAlpha const):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::createSwapchain const):

Canonical link: <a href="https://commits.webkit.org/300872@main">https://commits.webkit.org/300872@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4cf4ae1423ba9d3e833d905745c90784a15aa3d9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123561 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43276 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33972 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130289 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75703 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43999 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/51870 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/93931 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/62355 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126514 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35048 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110537 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74550 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34019 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73804 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104765 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/28919 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133014 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50512 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38454 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102427 "7 flakes 57 failures") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/50887 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106758 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102269 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26130 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47618 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/25853 "Passed tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47323 "Failed to compile WebKit") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50366 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56127 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/49840 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53187 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51515 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4cf4ae1423ba9d3e833d905745c90784a15aa3d9. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51507

djg reviewed (approved):
:shipit:

svillar committed:
[WebXR][OpenXR] Content not shown when asking for canvas without alpha layer
https://bugs.webkit.org/show_bug.cgi?id=299737

Reviewed by Dan Glastonbury.

Content was not rendered on screen because we were trying to
blit buffers with different pixel formats in case an alpha channel
was not requested. We were creating and OpenXR texture with GL_RGB
but WebGL's opaque framebuffer unconditionally uses GL_RGBA.

It is not possible to make the opaque framebuffer to use
RGB instead because it has always to import a texture with an
alpha channel because ANGLE does not support creating an
external texture from a framebuffer with RGB format
(see Source/ThirdParty/ANGLE/src/common/linux/dma_buf_utils.cpp).

So the only solution is to make OpenXR always create swapchains
with alpha channel and then when exporting the texture just
ignore it (by using DRM_FORMAT_XRGBA8888). This way WebGL
and OpenXR will use the same format and blitting will work.

* Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.cpp:
(WebKit::OpenXRLayer::exportOpenXRTextureGBM):
* Source/WebKit/UIProcess/XR/openxr/OpenXRSwapchain.cpp:
(WebKit::OpenXRSwapchain::create):
(WebKit::OpenXRSwapchain::OpenXRSwapchain):
* Source/WebKit/UIProcess/XR/openxr/OpenXRSwapchain.h:
(WebKit::OpenXRSwapchain::hasAlpha const):
* Source/WebKit/UIProcess/XR/openxr/PlatformXROpenXR.cpp:
(WebKit::OpenXRCoordinator::createSwapchain const):

Canonical link: https://commits.webkit.org/300872@main

webkit-commit-queue commented:
Committed 300872@main (60a2c8c774dc): <https://commits.webkit.org/300872@main>

Reviewed commits have been landed. Closing PR #51507 and removing active labels.
